### ISSUE TITLE
the rect in svg can not be rending when to canvas?

### ISSUE THREAD

**Issue by Undo101** [canvg/canvg#869]
In my svg, there are both text、rect  or path, and the text is above rect，the rect as a background-color block for the text. when i use "canvg.fromString" and "render" to make a canvas, text and path is ok, but rect not. can you tell how to fix it ? thanks

**Comment by gabelerner** [comment]
Paste your whole svg string please

**Comment by Undo101** [comment]
Like this:
<svg id="svg" class="iSvg_17x0g" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xml:space="preserve" height="307.05079605761944" width="375" viewBox="0 0 375 307.05079605761944" style="width: 375px; height: 307.051px; cursor: pointer;"><image xlink:href="/static/media/pic-test-wide.8c5dead7.jpg" x="0" y="0" height="307.05079605761944" width="375" transform="rotate(0, 187.5 153.52539802880972)" style="pointer-events: none;"></image><g><rect style="width:163px; height:35px; fill: #ff3d3d" rx="5" ry="5" x="106" y="136.02539802880972" id="textBgBlock1"></rect><text style="font-weight: bold; fill:#fff; pointer-events: none;" y="159.02539802880972" x="110" id="textContent1-0">正常秩序造成 v 分</text></g><path class="draw-path" stroke-linejoin="round" stroke-width="5" stroke="#FF3D3D" fill="#ffffff" stroke-opacity="1" fill-opacity="0" pointer-events="none" d="M 162.15625,212.55859375 L 171.04296875,219.328125 L 174.19921875,221.21875 L 177.78125,224.0859375 L 180.3046875,225.9765625 L 185.1171875,229.18359375 L 188.2734375,231.07421875 L 191.85546875,233.94140625 L 194.37890625,235.83203125 L 197.53515625,237.72265625 L 201.1171875,240.58984375 L 204.69921875,243.45703125 L 207.85546875,245.34765625 L 210.37890625,247.87109375 L 212.90234375,249.76171875 L 215.42578125,251.65234375 L 217.0625,253.2890625 L 218.44140625,254.20703125 L 219.8203125,255.125 L 220.5703125,255.875 L 221.3203125,256.25 L 222.0703125,257 L 222.359375,257.2890625 L 223.109375,257.6640625 L 223.3984375,257.953125 L 223.6875,258.2421875 L 223.9765625,258.53125 L 224.265625,258.8203125 L 224.5546875,259.109375 L 225.3046875,259.109375 L 226.0546875,259.484375 L 227.43359375,259.94140625 L 228.8125,260.3984375 L 230.99609375,261.48828125 L 232.375,262.40625 L 234.55859375,262.94921875 L 236.7421875,264.0390625 L 238.92578125,264.58203125 L 240.3046875,265.0390625 L 241.0546875,265.4140625 L 242.43359375,265.87109375 L 243.18359375,265.87109375 L 243.47265625,265.87109375 L 244.22265625,265.87109375 L 244.97265625,265.87109375 L 245.26171875,265.87109375 L 246.01171875,265.87109375 L 247.390625,264.94921875 L 248.140625,264.1953125 L 251.0078125,260.609375 L 252.64453125,258.96875 L 254.53515625,256.44140625 L 258.546875,251.625 L 262.12890625,248.75390625 L 265.3359375,243.9375 L 269.34765625,239.921875 L 273.359375,235.90625 L 277.37109375,231.08984375 L 281.3828125,226.2734375 L 285.82421875,220.05078125 L 290.6953125,212.25 L 295.13671875,206.02734375 L 297.02734375,202.8671875 L 300.578125,196.64453125 L 303.78515625,191.828125 L 305.93359375,187.5234375 L 308.08203125,183.21875 L 309.34375,180.05859375 L 311.234375,177.53125 L 312.49609375,174.37109375 L 314.38671875,171.84375 L 315.4765625,169.65625 L 317.3671875,167.12890625 L 318.45703125,164.94140625 L 320.34765625,162.4140625 L 321.265625,161.03125 L 322.35546875,158.84375 L 323.2734375,157.4609375 L 324.19140625,156.078125 L 324.94140625,155.32421875 L 325.859375,153.94140625 L 326.77734375,152.55859375 L 327.6953125,151.17578125 L 328.0703125,150.421875 L 328.98828125,149.0390625 L 329.90625,147.65625 L 330.65625,146.90234375 L 331.11328125,145.51953125 L 331.40234375,145.2265625 L 332.15234375,144.47265625 L 332.44140625,144.1796875 L 332.73046875,143.88671875 L 333.01953125,143.59375 L 333.01953125,143.33984375"></path></svg>

**Comment by Undo101** [comment]
Like this
"<svg id="svg" class="iSvg_17x0g" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xml:space="preserve" height="307.05079605761944" width="375" viewBox="0 0 375 307.05079605761944" style="width: 375px; height: 307.051px; cursor: pointer;"><image xlink:href="/static/media/pic-test-wide.8c5dead7.jpg" x="0" y="0" height="307.05079605761944" width="375" transform="rotate(0, 187.5 153.52539802880972)" style="pointer-events: none;"></image><g><rect style="width:163px; height:35px; fill: #ff3d3d" rx="5" ry="5" x="106" y="136.02539802880972" id="textBgBlock1"></rect><text style="font-weight: bold; fill:#fff; pointer-events: none;" y="159.02539802880972" x="110" id="textContent1-0">正常秩序造成 v 分</text></g><path class="draw-path" stroke-linejoin="round" stroke-width="5" stroke="#FF3D3D" fill="#ffffff" stroke-opacity="1" fill-opacity="0" pointer-events="none" d="M 162.15625,212.55859375 L 171.04296875,219.328125 L 174.19921875,221.21875 L 177.78125,224.0859375 L 180.3046875,225.9765625 L 185.1171875,229.18359375 L 188.2734375,231.07421875 L 191.85546875,233.94140625 L 194.37890625,235.83203125 L 197.53515625,237.72265625 L 201.1171875,240.58984375 L 204.69921875,243.45703125 L 207.85546875,245.34765625 L 210.37890625,247.87109375 L 212.90234375,249.76171875 L 215.42578125,251.65234375 L 217.0625,253.2890625 L 218.44140625,254.20703125 L 219.8203125,255.125 L 220.5703125,255.875 L 221.3203125,256.25 L 222.0703125,257 L 222.359375,257.2890625 L 223.109375,257.6640625 L 223.3984375,257.953125 L 223.6875,258.2421875 L 223.9765625,258.53125 L 224.265625,258.8203125 L 224.5546875,259.109375 L 225.3046875,259.109375 L 226.0546875,259.484375 L 227.43359375,259.94140625 L 228.8125,260.3984375 L 230.99609375,261.48828125 L 232.375,262.40625 L 234.55859375,262.94921875 L 236.7421875,264.0390625 L 238.92578125,264.58203125 L 240.3046875,265.0390625 L 241.0546875,265.4140625 L 242.43359375,265.87109375 L 243.18359375,265.87109375 L 243.47265625,265.87109375 L 244.22265625,265.87109375 L 244.97265625,265.87109375 L 245.26171875,265.87109375 L 246.01171875,265.87109375 L 247.390625,264.94921875 L 248.140625,264.1953125 L 251.0078125,260.609375 L 252.64453125,258.96875 L 254.53515625,256.44140625 L 258.546875,251.625 L 262.12890625,248.75390625 L 265.3359375,243.9375 L 269.34765625,239.921875 L 273.359375,235.90625 L 277.37109375,231.08984375 L 281.3828125,226.2734375 L 285.82421875,220.05078125 L 290.6953125,212.25 L 295.13671875,206.02734375 L 297.02734375,202.8671875 L 300.578125,196.64453125 L 303.78515625,191.828125 L 305.93359375,187.5234375 L 308.08203125,183.21875 L 309.34375,180.05859375 L 311.234375,177.53125 L 312.49609375,174.37109375 L 314.38671875,171.84375 L 315.4765625,169.65625 L 317.3671875,167.12890625 L 318.45703125,164.94140625 L 320.34765625,162.4140625 L 321.265625,161.03125 L 322.35546875,158.84375 L 323.2734375,157.4609375 L 324.19140625,156.078125 L 324.94140625,155.32421875 L 325.859375,153.94140625 L 326.77734375,152.55859375 L 327.6953125,151.17578125 L 328.0703125,150.421875 L 328.98828125,149.0390625 L 329.90625,147.65625 L 330.65625,146.90234375 L 331.11328125,145.51953125 L 331.40234375,145.2265625 L 332.15234375,144.47265625 L 332.44140625,144.1796875 L 332.73046875,143.88671875 L 333.01953125,143.59375 L 333.01953125,143.33984375"></path></svg>"

**Comment by Undo101** [comment]
Like this:(I use the "*" replace "<" to prevent code from being parsed)
*svg id="svg" class="iSvg_17x0g" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xml:space="preserve" height="307.05079605761944" width="375" viewBox="0 0 375 307.05079605761944" style="width: 375px; height: 307.051px; cursor: pointer;"><image xlink:href="/static/media/pic-test-wide.8c5dead7.jpg" x="0" y="0" height="307.05079605761944" width="375" transform="rotate(0, 187.5 153.52539802880972)" style="pointer-events: none;"></image><g><rect style="width:163px; height:35px; fill: #ff3d3d" rx="5" ry="5" x="106" y="136.02539802880972" id="textBgBlock1"></rect><text style="font-weight: bold; fill:#fff; pointer-events: none;" y="159.02539802880972" x="110" id="textContent1-0">正常秩序造成 v 分</text></g><path class="draw-path" stroke-linejoin="round" stroke-width="5" stroke="#FF3D3D" fill="#ffffff" stroke-opacity="1" fill-opacity="0" pointer-events="none" d="M 162.15625,212.55859375 L 171.04296875,219.328125 L 174.19921875,221.21875 L 177.78125,224.0859375 L 180.3046875,225.9765625 L 185.1171875,229.18359375 L 188.2734375,231.07421875 L 191.85546875,233.94140625 L 194.37890625,235.83203125 L 197.53515625,237.72265625 L 201.1171875,240.58984375 L 204.69921875,243.45703125 L 207.85546875,245.34765625 L 210.37890625,247.87109375 L 212.90234375,249.76171875 L 215.42578125,251.65234375 L 217.0625,253.2890625 L 218.44140625,254.20703125 L 219.8203125,255.125 L 220.5703125,255.875 L 221.3203125,256.25 L 222.0703125,257 L 222.359375,257.2890625 L 223.109375,257.6640625 L 223.3984375,257.953125 L 223.6875,258.2421875 L 223.9765625,258.53125 L 224.265625,258.8203125 L 224.5546875,259.109375 L 225.3046875,259.109375 L 226.0546875,259.484375 L 227.43359375,259.94140625 L 228.8125,260.3984375 L 230.99609375,261.48828125 L 232.375,262.40625 L 234.55859375,262.94921875 L 236.7421875,264.0390625 L 238.92578125,264.58203125 L 240.3046875,265.0390625 L 241.0546875,265.4140625 L 242.43359375,265.87109375 L 243.18359375,265.87109375 L 243.47265625,265.87109375 L 244.22265625,265.87109375 L 244.97265625,265.87109375 L 245.26171875,265.87109375 L 246.01171875,265.87109375 L 247.390625,264.94921875 L 248.140625,264.1953125 L 251.0078125,260.609375 L 252.64453125,258.96875 L 254.53515625,256.44140625 L 258.546875,251.625 L 262.12890625,248.75390625 L 265.3359375,243.9375 L 269.34765625,239.921875 L 273.359375,235.90625 L 277.37109375,231.08984375 L 281.3828125,226.2734375 L 285.82421875,220.05078125 L 290.6953125,212.25 L 295.13671875,206.02734375 L 297.02734375,202.8671875 L 300.578125,196.64453125 L 303.78515625,191.828125 L 305.93359375,187.5234375 L 308.08203125,183.21875 L 309.34375,180.05859375 L 311.234375,177.53125 L 312.49609375,174.37109375 L 314.38671875,171.84375 L 315.4765625,169.65625 L 317.3671875,167.12890625 L 318.45703125,164.94140625 L 320.34765625,162.4140625 L 321.265625,161.03125 L 322.35546875,158.84375 L 323.2734375,157.4609375 L 324.19140625,156.078125 L 324.94140625,155.32421875 L 325.859375,153.94140625 L 326.77734375,152.55859375 L 327.6953125,151.17578125 L 328.0703125,150.421875 L 328.98828125,149.0390625 L 329.90625,147.65625 L 330.65625,146.90234375 L 331.11328125,145.51953125 L 331.40234375,145.2265625 L 332.15234375,144.47265625 L 332.44140625,144.1796875 L 332.73046875,143.88671875 L 333.01953125,143.59375 L 333.01953125,143.33984375"></path>*/svg*

**Comment by gabelerner** [comment]
i can fix it for you @Undo101 , just paste it in any format.

you can also use three backticks (```) around your svg string for it to not parse weird.

**Comment by Undo101** [comment]
Like this:
`<svg id="svg" class="iSvg_17x0g" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xml:space="preserve" height="307.05079605761944" width="375" viewBox="0 0 375 307.05079605761944" style="width: 375px; height: 307.051px; cursor: pointer;"><image xlink:href="/static/media/pic-test-wide.8c5dead7.jpg" x="0" y="0" height="307.05079605761944" width="375" transform="rotate(0, 187.5 153.52539802880972)" style="pointer-events: none;"></image><g><rect style="width:163px; height:35px; fill: #ff3d3d" rx="5" ry="5" x="106" y="136.02539802880972" id="textBgBlock1"></rect><text style="font-weight: bold; fill:#fff; pointer-events: none;" y="159.02539802880972" x="110" id="textContent1-0">正常秩序造成 v 分</text></g><path class="draw-path" stroke-linejoin="round" stroke-width="5" stroke="#FF3D3D" fill="#ffffff" stroke-opacity="1" fill-opacity="0" pointer-events="none" d="M 162.15625,212.55859375 L 171.04296875,219.328125 L 174.19921875,221.21875 L 177.78125,224.0859375 L 180.3046875,225.9765625 L 185.1171875,229.18359375 L 188.2734375,231.07421875 L 191.85546875,233.94140625 L 194.37890625,235.83203125 L 197.53515625,237.72265625 L 201.1171875,240.58984375 L 204.69921875,243.45703125 L 207.85546875,245.34765625 L 210.37890625,247.87109375 L 212.90234375,249.76171875 L 215.42578125,251.65234375 L 217.0625,253.2890625 L 218.44140625,254.20703125 L 219.8203125,255.125 L 220.5703125,255.875 L 221.3203125,256.25 L 222.0703125,257 L 222.359375,257.2890625 L 223.109375,257.6640625 L 223.3984375,257.953125 L 223.6875,258.2421875 L 223.9765625,258.53125 L 224.265625,258.8203125 L 224.5546875,259.109375 L 225.3046875,259.109375 L 226.0546875,259.484375 L 227.43359375,259.94140625 L 228.8125,260.3984375 L 230.99609375,261.48828125 L 232.375,262.40625 L 234.55859375,262.94921875 L 236.7421875,264.0390625 L 238.92578125,264.58203125 L 240.3046875,265.0390625 L 241.0546875,265.4140625 L 242.43359375,265.87109375 L 243.18359375,265.87109375 L 243.47265625,265.87109375 L 244.22265625,265.87109375 L 244.97265625,265.87109375 L 245.26171875,265.87109375 L 246.01171875,265.87109375 L 247.390625,264.94921875 L 248.140625,264.1953125 L 251.0078125,260.609375 L 252.64453125,258.96875 L 254.53515625,256.44140625 L 258.546875,251.625 L 262.12890625,248.75390625 L 265.3359375,243.9375 L 269.34765625,239.921875 L 273.359375,235.90625 L 277.37109375,231.08984375 L 281.3828125,226.2734375 L 285.82421875,220.05078125 L 290.6953125,212.25 L 295.13671875,206.02734375 L 297.02734375,202.8671875 L 300.578125,196.64453125 L 303.78515625,191.828125 L 305.93359375,187.5234375 L 308.08203125,183.21875 L 309.34375,180.05859375 L 311.234375,177.53125 L 312.49609375,174.37109375 L 314.38671875,171.84375 L 315.4765625,169.65625 L 317.3671875,167.12890625 L 318.45703125,164.94140625 L 320.34765625,162.4140625 L 321.265625,161.03125 L 322.35546875,158.84375 L 323.2734375,157.4609375 L 324.19140625,156.078125 L 324.94140625,155.32421875 L 325.859375,153.94140625 L 326.77734375,152.55859375 L 327.6953125,151.17578125 L 328.0703125,150.421875 L 328.98828125,149.0390625 L 329.90625,147.65625 L 330.65625,146.90234375 L 331.11328125,145.51953125 L 331.40234375,145.2265625 L 332.15234375,144.47265625 L 332.44140625,144.1796875 L 332.73046875,143.88671875 L 333.01953125,143.59375 L 333.01953125,143.33984375"></path></svg>`

**Comment by Undo101** [comment]
For example, in svg:
![image](https://user-images.githubusercontent.com/26062774/71498766-951bf980-2898-11ea-8630-a3953f239373.png)
but after to canvas:
![image](https://user-images.githubusercontent.com/26062774/71498822-e62bed80-2898-11ea-9048-e9c8d6d6abbb.png)

**Comment by Undo101** [comment]
I have pinpointed the problem, when canvg render the rectElement, you get the size of rect from the attribute,but in my svg, i wrote the width and height in css.
Thank you for creating the canvg plugin, it really helped me save a lot of time.
But I think canvg should code a tool class to get the size or other attributes of the tag, and then serve each class like rectElement to get its related attributes. Maybe i can do this for canvg if i have time recently.

**Comment by Undo101** [comment]
I have pinpointed the problem, when canvg render the rectElement, you get the size of rect from the attribute, but in my svg, i wrote the width and height in css.
Thank you for creating the canvg plugin, it really helped me save a lot of time.
But I think canvg should code a tool class to get the size or other attributes of the tag, and then serve each class like rectElement to get its related attributes. Maybe i can do this for canvg if i have time recently.

**Comment by gabelerner** [comment]
@dangreen here's a simpler example:
```
<svg id="svg" class="iSvg_17x0g" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xml:space="preserve" height="307.05079605761944" width="375" viewBox="0 0 375 307.05079605761944" style="width: 375px; height: 307.051px; cursor: pointer;"><g><rect style="width:163px; height:35px; fill: #ff3d3d" rx="5" ry="5" x="106" y="136.02539802880972" id="textBgBlock1"></rect><text style="font-weight: bold; fill:#fff; pointer-events: none;" y="159.02539802880972" x="110" id="textContent1-0">正常秩序造成 v 分</text></g></svg>
```

basically we should read `width` and `height` from styles as well as attributes.

**Comment by dangreen** [comment]
@Undo101 please test fix in v3.0.4
